fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidVariantAttributes.java

```diff
@@ -0,0 +1,182 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.fluid;
+
+import java.util.Optional;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.fluid.FlowableFluid;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.sound.SoundEvents;
+import net.minecraft.text.Text;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.lookup.v1.custom.ApiProviderMap;
+import net.fabricmc.fabric.impl.transfer.TransferApiImpl;
+
+/**
+ * Common fluid variant attributes, accessible both client-side and server-side.
+ *
+ * <p><b>Experimental feature</b>, we reserve the right to remove or change it without further notice.
+ * The transfer API is a complex addition, and we want to be able to correct possible design mistakes.
+ */
+@ApiStatus.Experimental
+public class FluidVariantAttributes {
+	private static final ApiProviderMap<Fluid, FluidVariantAttributeHandler> HANDLERS = ApiProviderMap.create();
+	private static final FluidVariantAttributeHandler DEFAULT_HANDLER = new FluidVariantAttributeHandler() { };
+
+	/**
+	 * Register an attribute handler for the passed fluid.
+	 */
+	public static void register(Fluid fluid, FluidVariantAttributeHandler handler) {
+		if (HANDLERS.putIfAbsent(fluid, handler) != null) {
+			throw new IllegalArgumentException("Duplicate handler registration for fluid " + fluid);
+		}
+	}
+
+	/**
+	 * Return the attribute handler for the passed fluid, if available, and {@code null} otherwise.
+	 */
+	@Nullable
+	public static FluidVariantAttributeHandler getHandler(Fluid fluid) {
+		return HANDLERS.get(fluid);
+	}
+
+	/**
+	 * Return the attribute handler for the passed fluid, if available, or the default instance otherwise.
+	 */
+	public static FluidVariantAttributeHandler getHandlerOrDefault(Fluid fluid) {
+		FluidVariantAttributeHandler handler = HANDLERS.get(fluid);
+		return handler == null ? DEFAULT_HANDLER : handler;
+	}
+
+	/**
+	 * Return the name that should be used for the passed fluid variant.
+	 */
+	public static Text getName(FluidVariant variant) {
+		return getHandlerOrDefault(variant.getFluid()).getName(variant);
+	}
+
+	/**
+	 * Return the sound corresponding to a container of this fluid variant being filled if available,
+	 * or the default (water) filling sound otherwise.
+	 */
+	public static SoundEvent getFillSound(FluidVariant variant) {
+		return getHandlerOrDefault(variant.getFluid()).getFillSound(variant).orElse(SoundEvents.ITEM_BUCKET_FILL);
```

Comment:
Maybe fall back to the fluid's sound here?

Reply:
Yeah, I guess that should interact correctly with the mixin if it's done here.